geode-gfsh/build.gradle

```diff
@@ -25,6 +25,7 @@ dependencies {
 
     api(project(':geode-core'))
     api(project(':geode-common'))
+    api(project(':geode-wan'))
```

Comment:
I'm not sure this dependency should be added here. I think you can look into moving ReplicateRegionFunction.replicateRegion and all the methods it uses down into the Region hierarchy somewhere - probably DistributedRegion and PartitionedRegion depending on the Region type and some common methods in LocalRegion. I think that would eliminate this dependency from geode-gfsh. I haven't done any of this work, though, so I'm not sure the consequences. Your current replicateRegion method could do something like:

Your current replicateRegion method could be changed to something like:
```
private CliFunctionResult replicateRegion(FunctionContext<String[]> context, Region region,
    GatewaySender sender, long maxRate, int batchSize) {
  int replicatedEntries = 0;
  try {
    replicatedEntries = region.replicate(sender, maxRate, batchSize);
  } catch (InterruptedException e) {
    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
        "Operation canceled after having replicated " + replicatedEntries + " entries");
  } catch (BatchException70 e) {
    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
        "Error (" + e.getMessage() + ") in operation after having replicated "
            + replicatedEntries + " entries");
  }

  return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
      "Entries replicated: " + replicatedEntries);
}
```

Reply:
I see why you added that dependency. GatewaySenderBatchOp and GatewayAck are in geode-wan module. You won't be able to add it to geode-core either, so  I think you'll need to add an API to GatewaySenderEventDispatcher (which is in geode-core) for the sendBatch method. Something like:
```
sendBatch(Connection connection, PoolImpl senderPool, int batchId, int replicatedEntries, long maxRate, long startTime, List batch) 
  throws InterruptedException, BatchException70 {
```
Then implement it in GatewaySenderEventRemoteDispatcher and I guess throw UnsupportedOperationException in GatewaySenderEventCallbackDispatcher (unless we want to support sending all events in a region to an AEQ too).

Then call it from the GatewaySender's eventProcessor.

I'll have to look at this more tomorrow.

Reply:
Thanks for this suggestion. I have sent a new commit with this refactor to remove the dependency.